db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java

```diff
@@ -2178,6 +2196,17 @@ private void rememberBestCost(CostEstimate currentCost,int planType) throws Stan
         // Consider consolidating all these into nanoseconds in the future.
         if((bestCost.getEstimatedCost()/NANOS_TO_MILLIS)<timeLimit) {
             timeLimit=bestCost.getEstimatedCost()/NANOS_TO_MILLIS;
+
+            // Triggers and other prepared statements may have a very low
+            // cost estimate because the size of the involved trigger rows
+            // is not known up-front.  We don't want to quit query compilation
+            // early, resulting in a bad query plan or a plan which fails to
+            // pick up query hints, especially since the
+            // prepared statement does not need to be compiled on every
+            // invocation, just once in a while.  So, it is time well
+            // spent if it means we can compile a more optimal plan.
+            if (timeLimit < 2000 && isPreparedStatement())
```

Comment:
Nitpick, can you make it a static constant instead of a hardcoded number?

Reply:
Done.